dd-sdk-android-gradle-plugin/src/main/kotlin/com/datadog/gradle/plugin/DdAndroidGradlePlugin.kt

```diff
@@ -141,39 +139,21 @@ class DdAndroidGradlePlugin @Inject constructor(
                 variant
             )
             if (extensionConfiguration.checkProjectDependencies == SdkCheckLevel.NONE) {
-                return compileTask
-            }
-            // this will postpone the check until the actual compilation for the particular variant.
-            // by doing this we are avoiding pulling all the configurations for all the variants
-            // after build script is evaluated, which may be a heavy task for the big projects
-            return compileTask.doFirst {
-                val firstLevelModuleDependencies = variant.runtimeConfiguration
-                    .resolvedConfiguration
-                    .firstLevelModuleDependencies
-
-                if (!isDatadogDependencyPresent(firstLevelModuleDependencies)) {
-
-                    val sdkCheckLevel = extensionConfiguration.checkProjectDependencies
-                        ?: SdkCheckLevel.FAIL
-
-                    when (sdkCheckLevel) {
-                        SdkCheckLevel.FAIL -> {
-                            throw MissingSdkException(
-                                MISSING_DD_SDK_MESSAGE.format(variant.name)
-                            )
-                        }
-                        SdkCheckLevel.WARN -> {
-                            LOGGER.warn(MISSING_DD_SDK_MESSAGE.format(variant.name))
-                        }
-                        else -> {
-                            throw IllegalArgumentException(
-                                "This should never happen," +
-                                    " value=$sdkCheckLevel is not handled"
-                            )
-                        }
-                    }
-                }
+                return null
             }
+            val checkDepsTaskName = "checkSdkDeps${variant.name.capitalize()}"
+            val resolvedCheckDependencyFlag =
+                extensionConfiguration.checkProjectDependencies ?: SdkCheckLevel.FAIL
+            val checkDepsTask = target.tasks.register(
+                checkDepsTaskName,
+                DdCheckSdkDepsTask::class.java
+            ) {
+                it.configuration.set(variant.compileConfiguration)
+                it.sdkCheckLevel.set(resolvedCheckDependencyFlag)
+                it.variantName.set(variant.name)
+            }.get()
+            compileTask.finalizedBy(checkDepsTask)
```

Comment:
maybe `finalizedBy` supports `TaskProvider` as argument? In that case we could avoid calling `get` above and it would be nice for the lazy graph evaluation. Theoretically it should by looking on this https://docs.gradle.org/current/dsl/org.gradle.api.Task.html#org.gradle.api.Task.dependencies (`TaskProvider` is children of `Provider`)